src/serializer/visitors.js

```diff
@@ -94,6 +115,71 @@ export let ClosureRefReplacer = {
       }
     }
   },
+
+  // A few very simple dead code elimination helpers. Eventually these should be subsumed by the partial evaluators.
+  IfStatement: {
+    exit: function(path: BabelTraversePath, state: ClosureRefReplacerState) {
+      let node = path.node;
+      let testTruthiness = getLiteralTruthiness(node.test);
+      if (testTruthiness.known) {
+        if (testTruthiness.value) {
+          // Strictly speaking this is not safe: Annex B.3.4 allows FunctionDeclarations as the body of IfStatements in sloppy mode,
+          // which have weird hoisting behavior: `console.log(typeof f); if (true) function f(){} console.log(typeof f)` will print 'undefined', 'function', but
+          // `console.log(typeof f); function f(){} console.log(typeof f)` will print 'function', 'function'.
+          // However, Babylon can't parse these, so it doesn't come up.
+          path.replaceWith(node.consequent);
+        } else {
+          if (node.alternate !== null) {
+            path.replaceWith(node.alternate);
+          } else {
+            path.remove();
+          }
+        }
+      }
+    },
+  },
+
+  ConditionalExpression: {
+    exit: function(path: BabelTraversePath, state: ClosureRefReplacerState) {
+      let node = path.node;
+      let testTruthiness = getLiteralTruthiness(node.test);
+      if (testTruthiness.known) {
+        path.replaceWith(testTruthiness.value ? node.consequent : node.alternate);
+      }
+    },
+  },
+
+  LogicalExpression: {
+    exit: function(path: BabelTraversePath, state: ClosureRefReplacerState) {
+      let node = path.node;
+      let leftTruthiness = getLiteralTruthiness(node.left);
+      if (node.operator === "&&" && leftTruthiness.known) {
+        path.replaceWith(leftTruthiness.value ? node.right : node.left);
+      } else if (node.operator === "||" && leftTruthiness.known) {
+        path.replaceWith(leftTruthiness.value ? node.left : node.right);
+      }
+    },
+  },
+
+  WhileStatement: {
+    exit: function(path: BabelTraversePath, state: ClosureRefReplacerState) {
+      let node = path.node;
+      let testTruthiness = getLiteralTruthiness(node.test);
+      if (testTruthiness.known && !testTruthiness.value) {
+        path.remove();
+      }
+    },
+  },
+
+  DoWhileStatement: {
```

Comment:
For completeness/code coverage, please also add a testcase for "do...while".

Reply:
I'm not sure how. I don't know what assertion I can make about them, given the current test framework.

(I'd [love to have](https://github.com/facebook/prepack/pull/1061#discussion_r143315808) snapshot tests...)

Reply:
Maybe just use "does not contain: while" for now.

Reply:
Sure, done.